### PR TITLE
Make DPMultiheadAttention drop-in compatible with nn.MultiheadAttention

### LINKED ISSUE #123: DPMultiheadAttention is not drop-in compatible with nn.MultiheadAttention
## 🐛 Bug
DPMultiheadAttention is not drop-in compatible with nn.MultiheadAttention.

Opacus can only support certain "building block" layers (such as nn.Linear and nn.Conv2d), and layers written using those building blocks - in theory pretty much all of them, but in practice this leads to reimplementations. For example, while you can indeed write a LSTM using nn.Linear layers, the implementation we have in nn.LSTM doesn't do that (so that it can fuse ops and be faster). Whenever this happens, we reimplement the layer and make the update available in the Layers folder.

The aim of these reimplementations is that they all should be perfect drop-in replacements of their "normal PyTorch" counterparts, meaning that they should have exactly the same API, and exactly the same state_dict, so that users can train using the Opacus version, but then they can export the state_dict and load it for inference in a PyTorch environment that doesn't even have Opacus installed.

## Please reproduce using our [template Colab](https://colab.research.google.com/drive/1pCyN3_gN3WyhyQt8--eBPN5zVf2XZxiN?usp=sharing) and post here the link

Colab to repro this issue: https://colab.research.google.com/drive/1N4lqMmG21cROzx6qgJphTniMzcaX190C?usp=sharing

## To Reproduce

1. Instantiate a DPMultiheadAttention module
2. Instantiate a nn.MultiheadAttention with the same parameters
3. Their state_dicts should match!

## Additional context
There can be two reasons why this happens:

1. You have exactly the same matrices, but the names of the actual parameters are different due to how you organized your class.
2. You structured your layers differently so eg you have more/fewer matrices than the reference implementation

An example of 1 is the [DPLSTM](https://github.com/pytorch/opacus/blob/master/opacus/layers/dp_lstm.py). It's a big file, but essentially it has the same matrices as `nn.LSTM`, but they would be called something like `layers[0].ih` instead of `ih_0`. To fix this, we have an helper class named [ParamRenameModule](https://github.com/pytorch/opacus/blob/master/opacus/layers/param_rename.py) that allows you to specify a mapping. It's being used in DPLSTM, and it can help you here too.

In addition to this, here we also have Reason 2: we have a different number of matrices so we should align them. There are a few tricks that can help you here (this will teach you how `torch.nn.Module` works too!). Take a look at this code snippet to learn how PyTorch finds and registers parameters.

```python
class MyCoolModule(nn.Module):
    def __init__(self):
        super().__init__()  # starts the machinery to find submodules
        self.linear = nn.Linear(3, 2)  # this gets found and Torch will search for its params
        self.wrong_convs = [nn.Conv2d(i, 2*i) for i in range(5)]  # this DOES NOT get found!
        self.correct_convs = nn.ModuleList([nn.Conv2d(i, 2*i) for i in range(5)]) # This gets found!
        # Ditto with dicts and nn.ModuleDict

        # Note that all torch does is to recursively go inside and find nn.Parameters inside your 
        # submodules and calling self.register_parameter on those. So even if it doesn't find 
        # any params in self.wrong_convs, you CAN do this and it's exactly the same:
        for conv_module in self.wrong_convs:
            for pname, p in conv_module.named_parameters():
                registered_name = f"{conv_module}.{pname}"  # This is what nn.Module does, you can do whatever you want
                self.register_parameter(registered_name, p)
```

Remember: **Tensors are NOT Parameters!** Random tensors in a nn.Module won't get trained ever, they need to be boxed inside a nn.Parameter. Reference modules like `nn.Linear` do this for you already, so you don't generally notice this but this is how the sausage is made.

Here, you could use this behavior to your advantage by hiding things you don't want to be automatically registered and named by PyTorch, and instead manually registering the parameter with the name you want. This includes "tucking away" multiple matrices in a single one in a submodule, and hiding this submodule inside a list (list with single item) so that you can register it manually with a name that matches that of `nn.MultiheadAttention`.

### PR DESCRIPTION
**Pull request by Wei-1**
Summary: This PR is target to resolve #123 on GitHub by having an additional re-naming mechanism to match the `state_dict` structure of `nn.MultiheadAttention`.

Differential Revision: D40671870

### CODE REVIEW

**Comment by Wei-1** [comment]
You can test the edited part by running `pytest` within the `opacus` folder.
And you should see the following expected result:
```
============= 162 passed, 41 skipped, 4411 warnings in 243.18s (0:04:03) =============
```

File: opacus/layers/dp_multihead_attention.py [hunk 1]
```diff
@@ -329,3 +332,64 @@ def forward(
...
+
+        for hook in self._state_dict_hooks.values():
+            hook_result = hook(self, destination_alter, prefix, local_metadata)
+            if hook_result is not None:
+                destination_alter = hook_result
+
+        return destination_alter
+
+    def named_parameters(self, prefix: str = '', recurse: bool = True):
+        return self.state_dict(prefix = prefix).items()
```
Comment (ffuuugor) [inline opacus/layers/dp_multihead_attention.py:395]: Tbh, I would not change `named_parameters` - having this method not match the actual structure of the model would be very confusing and would lead to unexpected behaviour:
* `named_parameters` and `parameters` should return the same set of parameters
* sometimes you need to modify the output of these methods and you want them to point to the actual objects used in the model
* it would mess with calling model attributes directly

**Review by ffuuugor — COMMENTED** [review]
Thanks for the great work! Making this work is very intricate work and your efforts are hugely appreciated.

@alexandresablayrolles @karthikprasad  What are your thoughts on having `state_dict` not represent the actual physical structure of the model? Loading is not a problem, since `DPMultiheadAttention` can load both DP and vanilla state dictionaries.

On the one hand, it makes it easy to interoperate with vanilla `nn.MultiHeadAttention` - you can train a model with DP, save it's state_dict and then load this dict into a non-DP model.

On the other hand, I can potentially see issues with unexpected state_dict: keys not matching parameter names. That said, I don't see any immediate problems, but I might be missing something.

File: opacus/tests/dp_layers/dp_multihead_attention_test.py [hunk 2]
```diff
@@ -126,3 +130,37 @@ def test_attn(
...
+            atol=1e-5,
+            rtol=1e-3,
+            key_padding_mask=None,
+            need_weights=True,
+            attn_mask=None,
+        )
+
+        if "in_proj_bias" in dp_attn.state_dict():
+            dp_attn._register_state_dict_hook(remove_in_proj_bias_hook)
+            self.assertFalse("in_proj_bias" in dp_attn.state_dict())
```
[the remark this answers is not in the document]
Comment (Wei-1) [reply opacus/tests/dp_layers/dp_multihead_attention_test.py:166]: The `state_dict` function is copied from the original `nn.module`, and one function that is has is to have the capability to apply `hook` to modify the `state_dict`. This test is basically validating the `hook` can also work in the new `state_dict` function in our code.
The reason why I didn't directly use the `hook` function as the core method to modify `state_dict` is because I think it might cause confusion when people are trying to add, modify, or remove other `hook` in their network.

File: opacus/tests/dp_layers/dp_multihead_attention_test.py [hunk 3]
```diff
@@ -126,3 +130,37 @@ def test_attn(
             need_weights=True,
             attn_mask=None,
         )
+
+        attn.load_state_dict(dp_attn.state_dict())
```
[the remark this answers is not in the document]
Comment (Wei-1) [reply opacus/tests/dp_layers/dp_multihead_attention_test.py:134]: Make sense! I will do the modification tomorrow!

File: opacus/layers/dp_multihead_attention.py (hunk 1 above)
Comment (Wei-1) [reply opacus/layers/dp_multihead_attention.py:395]: I think that is a valid concern.
Currently, in `opacus/tests/dp_layers/common.py` line-220, we check if the two `named_parameters` can match.

```
        nn_params = dict(nn_module.named_parameters())
        dp_params = dict(dp_module.named_parameters())
```
How about we just change them into:
```
        nn_params = nn_module.state_dict()
        dp_params = dp_module.state_dict()
```
Since getting `named_parameters` and then change them back to `dict` is a bit weird as well.

**Comment by Wei-1** [comment]
> On the other hand, I can potentially see issues with unexpected state_dict: keys not matching parameter names. That said, I don't see any immediate problems, but I might be missing something.

As we try to cover the parameter naming logic in `nn.MultiheadAttention` and `DPMultiheadAttention`, I think the major problem might come with maintenance. Since the entire transformation logic is rule-based, things will most likely break when there are modifications in `nn.MultiheadAttention` or `DPMultiheadAttention` in their naming method.

File: opacus/tests/dp_layers/dp_multihead_attention_test.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ffuuugor) [reply opacus/tests/dp_layers/dp_multihead_attention_test.py:166]: Oh, I see, it is to check if state dict hooks still work, gotcha.  Can you maybe add a comment explaining this?

File: opacus/layers/dp_multihead_attention.py (hunk 1 above)
Comment (ffuuugor) [reply opacus/layers/dp_multihead_attention.py:395]: Yeah, that make sense

**Comment by Wei-1** [comment]
@ffuuugor Changes had been made to address the concerns. Please let me know if this make sense to you!

**Comment by ffuuugor** [comment]
Thanks for addressing the comments!
I believe this PR is close to landing, but we need to sort out one thing first. Due to some bug in how CircleCI works with phabricator, our main testing pipeline is not being triggered on this PR.

I can see that it won't pass the linter check. Please refer to our [Contributor's guide](https://github.com/pytorch/opacus/blob/main/CONTRIBUTING.md) and run isort/black/flake commands to check your code it formatted properly

**Comment by Wei-1** [comment]
Thanks, @ffuuugor! An update had been made to address the lint consistency with Black/ISort/Flake8.

**Comment by ffuuugor** [comment]
Hey
Thanks for taking care of this.
One last thing - sometimes `isort` give different recommendations depending on the version. I have mine set up exactly as CircleCI and it gives the following:

```diff
--- a/opacus/layers/dp_multihead_attention.py
+++ b/opacus/layers/dp_multihead_attention.py
@@ -14,14 +14,13 @@
 # limitations under the License.

 import warnings
+from collections import OrderedDict

 import torch
 import torch.nn as nn
 import torch.nn.functional as F
 from torch.nn.parameter import Parameter

-from collections import OrderedDict
-

 class SequenceBias(nn.Module):
```

Can you pls make the change to make the linter happy?

And I'm really sorry for back and forth on this. Tests not triggering for fbcode-exported PRs is painful and we're investigating.

**Review by ffuuugor — APPROVED** [review]
Thank you!

**Comment by Wei-1** [comment]
I just pushed a new version to address this issue! Please let me know if everything is good!
